lightly/api/api_workflow_upload_metadata.py

```diff
@@ -172,22 +178,34 @@ def upload_custom_metadata(self,
             raise ValueError("There exist image names in the custom metadata "
                              "without corresponding filenames on the server.")
 
-        if verbose:
-            # wrap samples in a progress bar
-            samples = tqdm.tqdm(samples)
-
-        for sample in samples:
-
-            metadata = filename_to_metadata[sample.file_name]
-
-            if metadata is not None:
-                # create a request to update the custom metadata of the sample
+        # retry upload if it times out
+        def upload_sample_metadata(request, sample):
+            return retry(
+                self._samples_api.update_sample_by_id,
+                request,
+                dataset_id=self.dataset_id,
+                sample_id=sample.id,
+            )
+
+        # limit number of concurrent requests
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            futures = []
+            for sample in samples:
+                metadata = filename_to_metadata[sample.file_name]
                 update_sample_request = SampleUpdateRequest(
-                    custom_meta_data=metadata
-                )
-                # send the request to the api
-                self._samples_api.update_sample_by_id(
-                    update_sample_request,
-                    self.dataset_id,
-                    sample.id
-                )
+                custom_meta_data=metadata
+            )
+                if metadata is not None:
+                    future = executor.submit(
+                        upload_sample_metadata,
+                        update_sample_request,
+                        sample,
+                    )
+                    futures.append(future)
+            
+            completed_futures = concurrent.futures.as_completed(futures)
```

Comment:
Could you somehow tell that this is an iterator?

Reply:
In general, some comments here would be nice.

Reply:
Added more comments